cpp/ql/src/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -0,0 +1,482 @@
+        /* This condition must determine the flow of control; that is, this
+         * node must be a top-level condition. */
+        this.controlsBlock(controlled, testIsTrue)
+        or
+        exists (IRGuardCondition ne
+        | this =  ne.(LogicalNotInstruction).getOperand() and
+          ne.controls(controlled, testIsTrue.booleanNot())) 
+    }
+
+    /** Holds if (determined by this guard) `left < right + k` evaluates to `isLessThan` if this expression evaluates to `testIsTrue`. */
+    cached predicate comparesLt(Instruction left, Instruction right, int k, boolean isLessThan, boolean testIsTrue) {
+        compares_lt(this, left, right, k, isLessThan, testIsTrue)
+    }
+
+    /** Holds if (determined by this guard) `left < right + k` must be `isLessThan` in `block`.
+        If `isLessThan = false` then this implies `left >= right + k`.  */
+    cached predicate ensuresLt(Instruction left, Instruction right, int k, IRBlock block, boolean isLessThan) {
+        exists(boolean testIsTrue |
+            compares_lt(this, left, right, k, isLessThan, testIsTrue) and this.controls(block, testIsTrue)
+        )
+    }
+
+    /** Holds if (determined by this guard) `left == right + k` evaluates to `areEqual` if this expression evaluates to `testIsTrue`. */
+    cached predicate comparesEq(Instruction left, Instruction right, int k, boolean areEqual, boolean testIsTrue) {
+        compares_eq(this, left, right, k, areEqual, testIsTrue)
+    }
+
+    /** Holds if (determined by this guard) `left == right + k` must be `areEqual` in `block`.
+        If `areEqual = false` then this implies `left != right + k`.  */
+    cached predicate ensuresEq(Instruction left, Instruction right, int k, IRBlock block, boolean areEqual) {
+        exists(boolean testIsTrue |
+            compares_eq(this, left, right, k, areEqual, testIsTrue) and this.controls(block, testIsTrue)
+        )
+    }
+
+    /**
+     * Holds if this condition controls `block`, meaning that `block` is only
+     * entered if the value of this condition is `testIsTrue`. This helper
+     * predicate does not necessarily hold for binary logical operations like
+     * `&&` and `||`. See the detailed explanation on predicate `controls`.
+     */
+    private predicate controlsBlock(IRBlock controlled, boolean testIsTrue) {
+        exists(IRBlock thisblock
+        | thisblock.getAnInstruction() = this
+        | exists(IRBlock succ, ConditionalBranchInstruction branch
+          | testIsTrue = true and succ.getFirstInstruction() = branch.getTrueSuccessor()
+            or
+            testIsTrue = false and succ.getFirstInstruction() = branch.getFalseSuccessor()
+          | branch.getCondition() = this and
+            succ.dominates(controlled) and
+            forall(IRBlock pred
+            | pred.getASuccessor() = succ
+            | pred = thisblock or succ.dominates(pred)))) // removed reachability condition - is that OK?
```

Comment:
In that case I think we need to add back the reachability requirement here. I suggest adding an `IRBlock.isReachableFromFunctionEntry()` predicate since this sort of side condition will be needed on most code that uses the dominance predicates that are already exposed on `IRBlock`.

Reply:
I'm adding that predicate and using it here - it causes one additional result in the test, which is actually a false positive because we don't have exception edges in the IR yet.